docs: add server side authorization documentation (#1315)
freeCodeCamp/chapter: CONTRIBUTING.md

[This](https://github.com/freeCodeCamp/chapter/contribute) is a good place to go if you are looking to get started.

> ## Authorizing GraphGQL requests with `@Authorized` decorator

Reviewer: I think it's better to be less specific, so it's more comfortable and easier for our future self to broaden it when necessary.

```suggestion
## Authorized GraphQL Requests
```

> ### Overview or roles

Reviewer: ```suggestion
### Current Roles
```

> Chapter uses role based permission system, with specific permission assigned to one or multiple roles, on one or more different _levels/scopes_).

Reviewer: ```suggestion
Chapter has a role based permission system, with specific permissions and different **scopes** assigned to each role.
```

* Instance role - per instance - every registered user has instance role.
* Chapter role - per chapter - chapter role is created for user when explicitly joining chapter, or performing on event actions requiring chapter role. Ie. Rsvping to event. Elevated chapter role for one chapter, doesn't impact chapter roles for different chapters.
* Event role - per event - chapter role is created for user when Rsvping to an event. Elevated event role (permissions) for a one event doesn't impact event roles for different events.

_Note: in MVP Event roles are not used._

> ### `@Authorized` decorator

Reviewer: ```suggestion
### `@Authorized` Decorator Usage in Chapter
```

Decorator is available as an import from `type-graphql`.

```ts
import { Authorized } from 'type-graphql';
```

> Adding it in resolver before specific request logic, wraps the request in the authorization logic. Only for authorized users logic of the request will be executed.

Reviewer: ```suggestion
Adding it in the resolver before specific request logic, will wrap the request in the authorization logic, which will execute only authorized users' logic of the request.
```

Author: The last part of the suggestion:
> which will execute only authorized users' logic of the request.

Changes the sense of the original sentence. The point is that when user isn't authorized, their request will not reach the actual code intended for handling that request (when user is authorized), but will be stopped earlier, at the authorization step.

Reviewer: What do you think of this?

```suggestion
Structuring your authorization logic, by adding a resolver before the specific request, will wrap that request. With this, the request will execute that logic only for authorized users
```

> By default this will mean only logged in user can be authorized.

Reviewer: ```suggestion
By default, this means only logged in users can be authorized.
```

```ts
  @Authorized(Permission.EventCreate)
  @Mutation(() => Event)
  async createEvent(
    @Arg('chapterId', () => Int) chapterId: number,
    @Arg('data') data: CreateEventInputs,
    @Ctx() ctx: Required<ResolverCtx>,
  ): Promise<Event | null> {
    /* ... */
  }
```

> To allow authorization logic connecting user roles on specific chapter or event, with their permissions for performing request, certain arguments are required in the request.

Reviewer: ```suggestion
To allow authorization logic connecting users' roles on a specific chapter or event, with their permissions for performing the request, certain arguments are required in the request.
```

| Role _level/scope_ | Required arguments in GraphQL request |
|:------------------:|:-------------------------------------:|
| Instance           | -                                     |
| Chapter            | `chapterId` or `eventId`              |
| Event              | `eventId`                             |

### Flowchart

```mermaid
flowchart LR;
    A["Authorize\n(logged in)"]
    I{Is allowed by\nInstance role?}
    CP{Is user banned\nfrom chapter?}
    C{Is allowed by\nChapter role?}
    E{Is allowed by\nEvent role?}
    Y[Authorized]
    N[Not authorized]

    A-->I;
    I-->|No| CP;
    I-->|Yes| Y;
    CP-->|Not banned| C;
    CP-->|Banned| N;
    C-->|No| E;
    C-->|Yes| Y;
    E-->|Yes| Y;
    E-->|No| N;
```

### Example

> Adding `@Authorize(Permissions.EventEdit)` before event editing logic will authorize user if one of the following applies:

Reviewer: ```suggestion
Adding `@Authorize(Permissions.EventEdit)` before an event -which edits the logic- will authorize the user if one of the following applies:
```

Author: Could you take another look here? Suggestion is also altering the sense of original sentence.

Reviewer: Sure, that what I originally thought was going one 👍. Hopefully I will think of better suggestion

Reviewer: I had an issue with `event editing logic` because it feels like it's messing something. But that maybe me overthinking, so feel free to discard these suggestions if it's harmful to the overall document

* Their Instance role has `Permissions.EventEdit` permission.
* User is not banned on the chapter, on which edited event is, and for that chapter their Chapter role has `Permission.EventEdit` permission.
* Their Event role, for the edited event, has `Permission.EventEdit` permission.

> Otherwise request will be rejected.

Reviewer: ```suggestion
Otherwise, the request will be rejected.
```
